src/lib/storage/StorageApi.ts

```diff
@@ -0,0 +1,269 @@
+   */
+  async deleteBucket(
+    id: string
+  ): Promise<{ data: { message: string } | null; error: Error | null }> {
+    try {
+      const data = await remove(`${this.url}/bucket/${id}`, {}, { headers: this.headers })
+      return { data, error: null }
+    } catch (error) {
+      return { data: null, error }
+    }
+  }
+
+  /**
+   * Upload a file
+   * @param path the relative file path
+   * @param file the File content
+   */
+  async uploadFile(
+    path: string,
+    file: File
+  ): Promise<{ data: { message: string } | null; error: Error | null }> {
+    try {
+      if (!isBrowser()) throw new Error('No browser detected.')
+
+      const formData = new FormData()
+      formData.append('cacheControl', '3600')
+      formData.append('', file, file.name)
+
+      const res = await fetch(`${this.url}/object/${path}`, {
+        method: 'POST',
+        body: formData,
+        headers: { ...this.headers },
+      })
+
+      if (res.ok) {
+        const data = await res.json()
+        return { data, error: null }
+      } else {
+        const error = await res.json()
+        return { data: null, error }
+      }
+    } catch (error) {
+      return { data: null, error }
+    }
+  }
+
+  /**
+   * Update a file
+   * @param path the relative file path
+   * @param file the File content
+   */
+  async updateFile(
+    path: string,
+    file: File
+  ): Promise<{ data: { Key: string } | null; error: Error | null }> {
+    try {
+      if (!isBrowser()) throw new Error('No browser detected.')
+
+      const formData = new FormData()
+      formData.append('cacheControl', '3600')
+      formData.append('', file, file.name)
+
+      const res = await fetch(`${this.url}/object/${path}`, {
+        method: 'PUT',
+        body: formData,
+        headers: { ...this.headers },
+      })
+
+      if (res.ok) {
+        const data = await res.json()
+        return { data, error: null }
+      } else {
+        const error = await res.json()
+        return { data: null, error }
+      }
+    } catch (error) {
+      return { data: null, error }
+    }
+  }
+
+  /**
+   * Rename a file
+   * @param bucketName
+   * @param fromPath original relative file path
+   * @param toPath the new relative file path
+   */
+  async renameFile(
+    bucketName: string,
+    fromPath: string,
+    toPath: string
+  ): Promise<{ data: { message: string } | null; error: Error | null }> {
+    try {
+      const data = await post(
+        `${this.url}/object/rename`,
+        { bucketName, sourceKey: fromPath, destinationKey: toPath },
+        { headers: this.headers }
+      )
+      return { data, error: null }
+    } catch (error) {
+      return { data: null, error }
+    }
+  }
+
+  /**
+   * Create signed url to download file
+   * @param path the relative file path to be downloaded
+   * @param expiresIn seconds until the signed URL expires
+   */
+  async createSignedUrl(
+    path: string,
+    expiresIn: number
```

Comment:
I wonder if this should be set on the Server? Unsure, but I would imagine we don't want users to generating a tonne of signed URLs which last 1 year?

Reply:
I think it is okay to be set on the client. If the client has read access to the object, he can share that object to anyone. There are usecases for both short and long expiry times. Short expiry is useful for generating a temporary download link to the object and long times are useful say if you want to embed the resource on your website markup for example `<img src={{supabaseURL}} />`.